### PR TITLE
Add compare to Ord

### LINKED ISSUE #1634: Add `compare` to Ord and Rewrite Spaceship to use it

### PR DESCRIPTION
**Pull request by esbenbjerre**
- [x] Add `compare` to `Order`
- [x] Make `<=>` an alias for `compare`
- [x] Add `Order` constraint to `RedBlackTree`
- [x] Add `Order` instances for n-tuples with n in [2..10]
- [x] Rename `Ordering` to `Comparison` and move it to root
- [x] Rename `Equal` to `EqualTo`
- [x] Add methods `toInt`, `fromInt`, `min` and `max` to the `Order` namespace
- [x] Add tests for above methods

Fixes #1634.

### CODE REVIEW

File: main/src/library/Ord.flix [hunk 1]
```diff
@@ -120,5 +170,493 @@ instance Ord[BigInt] {
...
+    def greaterEqual(x: (a1, a2, a3, a4, a5, a6, a7, a8, a9, a10), y: (a1, a2, a3, a4, a5, a6, a7, a8, a9, a10)): Bool = match (x <=> y) {
+        case GreaterThan => true
+        case Equal => true
+        case _ => false
+    }
+
+    ///
+    /// Compares `t1` and `t2` lexicographically.
+    ///
+    def compare(t1: (a1, a2, a3, a4, a5, a6, a7, a8, a9, a10), t2: (a1, a2, a3, a4, a5, a6, a7, a8, a9, a10)): Ordering =
```
[the remark this answers is not in the document]
Comment (magnus-madsen) [reply main/src/library/Ord.flix:617]: And maybe instead of `|>` and `lazyCombine` we can invent one operator (with some fancy symbol) that does both.

**Review by magnus-madsen — COMMENTED** [review]
Big PR. I have a few changes to make it even bigger (!):

- [ ] Rename `Ordering` to `Comparison` and move it to `Comparison.flix` in the root namespace and folder.
- [ ] Rename `Equal` to `EqualTo` (I feel this flows better with `LessThan`, `EqualTo`, `GreaterThan`).
- [ ] Rename `Ord` to `Order`.
- [ ] Add the following methods to the Order namespace: 
  - `toInt(c: Comparison): Int32`
  - `fromInt(i: Int32): Comparison`
  - `min(x: a, y: a): a`
  - `max(x: a, y: a): a`
  - Add a few tests for each. Nothing extravagant.

File: main/src/library/Set.flix [hunk 2]
```diff
@@ -46,6 +47,30 @@ instance Eq[Set[a]] with[a : Eq] {
...
+    def greater(x: Set[a], y: Set[a]): Bool = match (x <=> y) {
+        case GreaterThan => true
+        case _ => false
+    }
+    def greaterEqual(x: Set[a], y: Set[a]): Bool = match (x <=> y) {
+        case GreaterThan => true
+        case Equal => true
+        case _ => false
+    }
+    pub def compare(xs: Set[a], ys: Set[a]): Ordering = Set.toList(xs) <=> Set.toList(ys)
```
[the remark this answers is not in the document]
Comment (esbenbjerre) [reply main/src/library/Set.flix:70]: It's a relic from when I implemented `RedBlackTree`. The main issue is that there can be multiple trees representing the same set. I have not yet stumbled upon an elegant algorithm for comparing two trees.

File: main/test/flix/Test.Exp.Binary.Spaceship.flix [hunk 3]
```diff
@@ -1,231 +1,256 @@
...
-    @test
-    def testBinarySpaceship31(): Bool = "hello" <=> "world" < 0
+///    @test
+///    def testBinarySpaceship30(): Bool = "hello" <=> "hello" == Equal
+///
+///    @test
+///    def testBinarySpaceship31(): Bool = "hello" <=> "world" == LessThan
+///
+///    @test
+///    def testBinarySpaceship32(): Bool = "world" <=> "hello" == GreaterThan
```
[the remark this answers is not in the document]
Comment (esbenbjerre) [reply main/test/flix/Test.Exp.Binary.Spaceship.flix:100]: An oversight :)

File: main/src/library/Set.flix (hunk 2 above)
[the remark this answers is not in the document]
Comment (magnus-madsen) [reply main/src/library/Set.flix:70]: What about:

```
if (xs subsetOf ys) => LessThan
else if (ys subsetOf xs) => GreaterThan
else EqualTo
```

Does that make sense?

[the remark this answers is not in the document]
Comment (magnus-madsen) [reply main/src/library/Set.flix:70]: I guess the question is: What ordering are you expecting?

File: main/src/library/Ord.flix (hunk 1 above)
[the remark this answers is not in the document]
Comment (esbenbjerre) [reply main/src/library/Ord.flix:617]: > 🤣
> I mean, I am glad it works, but wow!

It wasn't that bad, I made a Python script to generate the code. The hardest part was to look at it.





> Do you think something like the following could be made to work (better names needed):

Yes I get the idea. I think that would be nice. I'll take a look at it.

File: main/src/library/Set.flix (hunk 2 above)
[the remark this answers is not in the document]
Comment (esbenbjerre) [reply main/src/library/Set.flix:70]: > I guess the question is: What ordering are you expecting?

I think I would expect a lexicographic order. But I can definitely accept your proposal.

**Comment by esbenbjerre** [comment]
> * [ ]  Rename `Ordering` to `Comparison` and move it to `Comparison.flix` in the root namespace and folder.

Root folder as in /main/src/library?

> * [ ]  Add the following methods to the Order namespace:

As in add them to the typeclass and implement identical definitions for all instances?

File: main/src/library/Set.flix (hunk 2 above)
[the remark this answers is not in the document]
Comment (esbenbjerre) [reply main/src/library/Set.flix]: Now I remember why it is the way it is. If we define the ordering for `Set` as you propose we break nested sets (and maps).

[the remark this answers is not in the document]
Comment (esbenbjerre) [reply main/src/library/Set.flix]: This works.
```
pub def compare(xs: Set[a], ys: Set[a]): Comparison = match (Set.isSubsetOf(xs, ys), Set.isSubsetOf(ys, xs)) {
    case (true, true) => EqualTo
    case (false, true) => LessThan
    case _ => GreaterThan
}
```

File: main/src/library/Ord.flix (hunk 1 above)
[the remark this answers is not in the document]
Comment (esbenbjerre) [reply main/src/library/Ord.flix]: I think this PR is ready for merge modulo your proposal of the lazy combinator. Maybe we can add it afterwards?

**Comment by magnus-madsen** [comment]
> > * [ ]   Rename `Ordering` to `Comparison` and move it to `Comparison.flix` in the root namespace and folder.
> 
> Root folder as in /main/src/library?

Yes

File: main/src/library/Ord.flix (hunk 1 above)
[the remark this answers is not in the document]
Comment (magnus-madsen) [reply main/src/library/Ord.flix]: We could, will you open a ticket?

File: main/src/library/Set.flix (hunk 2 above)
[the remark this answers is not in the document]
Comment (magnus-madsen) [reply main/src/library/Set.flix]: OK; the order I suggested does not make sense; its only a partial order.

In fact, I think I don't want *any* Order instances for any set or map (!) So please go a head a remove those (and any test cases).

**Comment by magnus-madsen** [comment]
I think this is ready other than that I would like the `Order` instances for Set, MutSet, Map, and MutMap deleted.

File: main/src/library/Set.flix (hunk 2 above)
[the remark this answers is not in the document]
Comment (esbenbjerre) [reply main/src/library/Set.flix]: > OK; the order I suggested does not make sense; its only a partial order.
> 
> In fact, I think I don't want _any_ Order instances for any set or map (!) So please go a head a remove those (and any test cases).

Fair enough. What do we do about `flatten` and `subsets`? Remove them too?

File: main/src/library/Ord.flix (hunk 1 above)
[the remark this answers is not in the document]
Comment (esbenbjerre) [reply main/src/library/Ord.flix]: > We could, will you open a ticket?

Yes, I've opened #1638.

File: main/src/library/Set.flix (hunk 2 above)
[the remark this answers is not in the document]
Comment (magnus-madsen) [reply main/src/library/Set.flix]: OK, I see where this is going. Hmm.... I will merge this now and create a separate ticket for ordering on Sets / Maps.

**Pull request merged**